### PR TITLE
Gc partition state

### PR DESCRIPTION
**Pull request by jiangxinmeng1**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #11613

## What this PR does / why we need it:

现在CN不会gc logtail，如果CN会一直订阅一张表，这张表的logtail会无限增长。本PR会GC CN 内存中的logtail：

1. GC partition state中的block及index

   每当有block被删除，CN会收到metadata相关的logtail。CN只是逻辑删除这些block，它们依然占用内存。这个PR里，等block被逻辑删除超过`1h`，就会物理删除它们。

2. 增加后台任务partitionStateGCTicker

   partitionStateGCTicker每`20min`扫描一次CN中所有的partition state，GC block及其索引。

3. 取block list的时候检查timestamp

    新的事务带着SnapshotTS向partition state取block list，SnapshotTS控制可见性。如果SnapshotTS小于gc过的TS，partition state中可能有些block，本该被这个事务看见，但已经被GC了。CN不能服务这样的事务，会返回错误`TxnStale`。

### CODE REVIEW

**Review by zhangxu19830126 — APPROVED** [review]

File: pkg/vm/engine/disttae/logtailreplay/partition.go [hunk 1]
```diff
@@ -119,3 +119,14 @@ func (p *Partition) ConsumeCheckpoints(
 
 	return
 }
+
+func (p *Partition) Truncate(ctx context.Context, ids [2]uint64, ts types.TS) error {
+	err := p.Lock(ctx)
+	if err != nil {
+		p.Unlock()
```
Comment (reusee) [inline pkg/vm/engine/disttae/logtailreplay/partition.go:126]: Unlock should only be called when Lock returns nil.

File: pkg/vm/engine/disttae/logtailreplay/partition.go [hunk 2]
```diff
@@ -119,3 +119,14 @@ func (p *Partition) ConsumeCheckpoints(
...
+		return err
+	}
```
Comment (reusee) [inline pkg/vm/engine/disttae/logtailreplay/partition.go:128]: should put `defer p.Unlock()` here.

File: pkg/vm/engine/disttae/logtailreplay/partition_state.go [hunk 3]
```diff
@@ -68,6 +69,10 @@ type PartitionState struct {
 	// should have been in the Partition structure, but doing that requires much more codes changes
 	// so just put it here.
 	shared *sharedStates
+
+	// blocks deleted before minTS is hard deleted.
+	// partition state can't serve txn with snapshotTS less than minTS
+	minTS atomic.Pointer[types.TS]
```
Comment (reusee) [inline pkg/vm/engine/disttae/logtailreplay/partition_state.go:75]: this field should be types.TS and never be modified. The PartitionState is immutable.

File: pkg/vm/engine/disttae/logtailreplay/partition_state.go [hunk 4]
```diff
@@ -660,3 +666,60 @@ func (p *PartitionState) consumeCheckpoints(
 	p.checkpoints = p.checkpoints[:0]
 	return nil
 }
+
+func (p *PartitionState) truncate(ids [2]uint64, ts types.TS) {
```
Comment (reusee) [inline pkg/vm/engine/disttae/logtailreplay/partition_state.go:670]: This method should be moved to Partition. And never modify a PartitionState, copy and modify instead. See Partition.ConsumeCheckpoints and Partition.Mutate for how to do modifications.

**Review by reusee — CHANGES_REQUESTED** [review]

File: pkg/vm/engine/disttae/logtailreplay/blocks_iter.go [hunk 5]
```diff
@@ -31,13 +32,16 @@ type blocksIter struct {
 	firstCalled bool
 }
 
-func (p *PartitionState) NewBlocksIter(ts types.TS) *blocksIter {
+func (p *PartitionState) NewBlocksIter(ts types.TS, tid uint64) (*blocksIter, error) {
```
Comment (reusee) [inline pkg/vm/engine/disttae/logtailreplay/blocks_iter.go:35]: if the tid param is not used, delete it.

**Review by reusee — CHANGES_REQUESTED** [review]

**Review by reusee — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Pull request merged**